README.md
# firtz


*firtz - podcast publisher*
*Version 2.9.1 (Alpha)*

## Über firtz (RC2.9.1)

<pre>
Du hast offensichtlich den Drang, entweder diesen Kram hier auszuprobieren,
oder einen Podcast aus der Taufe zu heben. Was auch immer Dich dazu bewogen
hat firtz auszuprobieren: Gute Idee :)

Um einen Podcast zu veröffentlichen, reichen im Grunde zwei Dinge:
- Audiodateien im Web und irgendwo eine XML-Datei, die den Feed beschreibt.
- Alles andere ist Beiwerk, das mehr oder minder nötig und arbeitsintensiv ist.

Warum aber an vielen Stellen so viel Arbeit machen, so viel Ballast herumschleppen?
Wieso ein Wordpress, wenn man alle zwei Monate mal was veröffentlicht? Wieso eine
SQL-Datenbank für fünf Artikel in einem Jahr? Wieso sich den Kopf zerbrechen über
die Installation diverser Plugins, wenn es ein paar KByte XML tun?

Der firtz soll - wo möglich - dabei helfen, Podcasts den Druck zu nehmen, sich
um technische Details zu kümmern. Deshalb also firtz. Die 0.1 war das Ergebnis
von knapp einskommafünf Tagen "Arbeit", die ich eigentlich nur gemacht habe,
um mal wieder irgendwas zusammen zu schrauben, das halbwegs funktioniert.

Wie funktioniert das aber nun? Über Konfigurationsdateien. Eine für den Feed mit
allgemeinen Daten und einer Datei für jede Episode mit den Details eben jener.
Wer es ganz einfach haben möchte, verbündet sich mit Auphonic und überlässt alle
Daten, die dort anfallen dem firtz, der dann daraus einen oder mehrere Feeds erzeugt.
Einfacher wird's nur noch, wenn Auphonic selbst sowas anbietet ;)

-- eazy, 2013
</pre>

Change: M.C., 2019

## Voraussetzungen

Aber fangen wir vorne an. Was brauchst Du? Du brauchst irgendwo Platz im Web.
Das ist dann doch noch nötig. Dieser Platz im Web muss im Grunde nur eine
Bedingung erfüllen: **php ab Version 5.4.0** muss ausgeführt werden können.
Besser ist **php mit Version 7.3.x**. Datenbanken und anderen Kram braucht
es nicht, allerdings muss der ausführende Nutzer des Webservers in der Lage,
Verzeichnisse zu erzeugen und Dateien in diese hineinzuschreiben.

Sollte der firtz in einem Unterverzeichnis der genutzten Domain betrieben werden,
musst Du noch die .htaccess korrigieren und die RewriteBase von:
`RewriteBase /` auf `RewriteBase /UNTERORDNER` zu ändern.

Es gibt übrigens hier und dort Probleme mit dem Schreiben in das temporäre
Verzeichnis. Je nach Webhoster kann es nötig und hilfreich sein, den Handler
für php auf (fast-)cgi zu stellen. In der .htaccess geht das z.B. mit dieser
Zeile:  `AddHandler php5-cgi .php`

Das sollte allerdings wirklich nur passieren, wenn der firtz diesbezüglich
Fehlermeldungen auswirft.

Change: M.C., 2020

## Los geht's

Du besorgst Dir das [firtz-Archiv](https://github.com/Firtz-Designs/firtz3.0),
entpackst es in den Ordner, in dem Du das auf dem Webserver haben willst und
bewegst Dich darauf hin in den Unterordner `feeds/`. In diesem findest Du
ein Demo-Verzeichnis. Der Name dieses Ordners, ist der zukünftige Name
Deines Feeds. Im Demo-Ordner sind vier Typen von Dateien: `feed.cfg`,
`001.epi`, `001-banner.jpg` und `001-cover.png`. Letzere sind Banner und
(wenn Du es bevorzugst) Episoden Bild für jede Episode extra.

Die `feed.cfg` ist Deine Feed Datei mit allen Grunddaten die dein Podcast
benötigt. Die `001.epi` ist die Datei für eine Episode. Du solltest darauf
achten, das jede Datei **möglichst mit Zahl** abgebildet wird. Du kann auch
etwa so aussehen: `001-TolleEpisode.epi`


Das nicht syntaktisch zu nennende Prinzip dieser Dateien ist identisch:

`#: ist ein Kommentar nicht den ":" vergessen!`

Attribute werden wie folgt notiert:

<pre>
attribute:
value
</pre>

Achtet bitte darauf, dass die Attribut-Zeile alleine steht. Leerzeilen
werden ignoriert, es sei denn, es geht um eine Textfeld wie summary.
An's Ende der Konfiguration kannst Du in einer neuen Zeile ein

`---end---`

setzen, alles dahinter wird ignoriert. Du kannst da Notizen oder Müll
unterbringen. Ich werde das hier im Moment noch nicht weiter ausbreiten.
Schau in die Dateien und spiel damit rum.

**Wichtig ist, dass alle Konfigurationsdateien UTF8 sind, denn sonst gibt's
Müsli im Feed!**

Solltest Du [Auphonic](http://auphonic.com/) nutzen, bist Du ganz fein raus.
Aktiviere in den Ausgangsdateien dort die Production-Description,
sorge dafür, dass firtz sie im Dateisystem findet und schon erstellen
sich die Folgen fast von selbst anhand der von Auphonic weitergereichten
Metadaten. Weitere Informationen dazu finden sich in der weiterführenden
Dokumentation.

Nehmen wir an, Du hast das soweit fertig. Wie erreichst du den Feed?
Angenommen der URL zu Deinem Webserver wäre `http://tollerneuerpodcast.de/`.
Der Feed heißt ja noch `demo`, also wäre der URL zum RSS2-Feed:
`http://tollerneuerpodcast.de/demo`

Wenn Du mehrere Audioformate nutzt, kannst Du auch die URL je Format nutzen:
`http://tollerneuerpodcast.de/demo/mp3`.

Eine Webseite gibt es als Bonus automatisch dazu. Die kannst Du für dieses
Beispiel dann so erreichen: `http://tollerneuerpodcast.de/demo/show`

Für jede Episode gibt es einen eigenen [Podlove Webplayer](https://github.com/gerritvanaaken/podlove-web-player).

Einzelne Episoden sind auch verlinkbar, indem an den Pfad der Slug der
Episode (praktisch der Dateiname ohne Endung) angehangen wird:
`http://tollerneuerpodcst.de/demo/show/001`.

## Das Theme und Layout: QuorX 3

QuorX ist einfach veränderbar. Du findest die Einstellungen unter:
`/templates/default/template.cfg`. Ich habe Euch schon beschrieben, welche
Einstellungen Ihr machen könnt, um Teile des Designs zu verändern.

Hier also ein paar Beispiele.

### Breite des Themes (Box/Full)

**layout: box**

<img src="https://raw.githubusercontent.com/Firtz-Designs/firtz3.0/master/doc/img/screen1.png">

**layout: full**

<img src="https://raw.githubusercontent.com/Firtz-Designs/firtz3.0/master/doc/img/screen2.png">


### Farbe des Themes

Ich habe euch schon einige Farbthemen als Beispiele beigelegt.
Selbsterklärend, könnt Ihr so euer eigenes Farbschema entwerfen.

<pre>
color #31424b     (Player)
light #81b55e     (Buttons und Icons)
dark  #2c3b43     (Links & Hintergrund)
</pre>

<img src="https://raw.githubusercontent.com/Firtz-Designs/firtz3.0/master/doc/img/screen3.png">

Bitte achte darauf: Es sind diese Farbangaben auch in der feed.cfg
unter: `/feeds/<demo>/feed.cfg` möglich. Sind hier Angaben hinterlegt,
dann gelten diese immer als gesetzt.

Es gibt noch viele weitere Angaben zu entdecken die Du setzen kannst,
um dem Quorx3 Theme einen eigenen Look zu verpassen.

Viel Spaß dabei!

-- Stand: 08.06.2020 by McCouman


## Twitter Kontakte

- Developer: [@eazy](https://twitter.com/eazyliving)
- Developer: [@mccouman](https://twitter.com/mccouman)


## Links

- Theme: [QuorX 3](https://github.com/Firtz-Designs/QuorX-III)
- Framework: [fat free framework](https://github.com/bcosca/fatfree)
- Componense: [Podlove Webplayer](https://github.com/gerritvanaaken/podlove-web-player)


CHANGELOG.md
Changelog:

RC 3.0
- i.A.

2.9.1 Fix #1
- fixed xml summary output with summary_xml
- fixed rendering for set: home.html,home-sidebar.html
- change: classes/episode.php for summary output
- change: new var "summary_xml"; for correct feed output in tag itunes:summary
- change: take out strip_tags for better performance
- new: summary is now enabled for markdown
- update demo 001.epi file

2.9 Extensions and QuorX 3 changes
- fixed PHP 7.2 and 7.3 changes (string to array)
- extension: disqus
  - fixed js,
  - hooks
  - new layout
  - isqus create infobox in QuorX 3 style
- pages: Netiquette for disqus link
- fixed full width for webplayer
- extension: contributors
  - delete google+
  - create mastodon
- QuorX3
  - navi: social links as blank target
  - navi: new social icon mastodon
  - js icon callback with mastodon
  - mastodon style icon
  - delete styles for webplayer
  - delete styles for subscribe button
  - delete middle container in set
  - change dict for translation (de)
  - delete old admin translation

2.8 PHP fixes for 7.x and new Attributes + QuorX 3
- fixed PHP 7.0 (eazy)
- fixed PHP 7.3 (mccouman)
- fixed webplayer extension
- new QuorX 3 Skin
  - style guides
  - theme functions
  - more hooks
  - contributors view (guest,podcaster)
- new meta types: feedattr_default
  - 'subtitle',         # Update for podlove webplayer
  - 'street',           # Impressum (required fields)
  - 'city',             # Impressum (required fields)
  - 'country',          # Impressum (required fields)
  - 'phone',            # Impressum (required fields)
  - 'fax',              # Impressum (required fields)
  - 'tax',              # Impressum (required fields)
  - 'image',            # QuorX 3 Skin style (required fields)
  - 'imageLicense',     # QuorX 3 Skin style (optional)
  - 'imageCreator',     # QuorX 3 Skin style (optional)
  - 'imageURL',         # QuorX 3 Skin style (optional)
  - 'supporter'         # Sidebar project support (optional)
- new meta types: itemattr
  - 'subtitle',         # Update podlove webplayer / xml
  - 'summary',          # Update for podlove webplayer
  - 'banner',           # QuorX 3 Skin style (optional)
  - 'bannerLicense',    # QuorX 3 Skin style (optional)
  - 'bannerPage',       # QuorX 3 Skin style (optional)
  - 'bannerURL',        # QuorX 3 Skin style (optional)
  - 'image',            # QuorX 3 Skin style (optional)
  - 'intro',            # QuorX 3 Skin style (optional)
  - 'outro',            # QuorX 3 Skin style (optional)

2.1 Beta 2
- fixed site.html
- fixed extension (description as JSON output)
- fix share json
- fix feed class
- remove pwp folder (obsolet!)

2.1 firtz + podlove webplayer 4
- remove search extension and include in firtz (redirect)
- new extension "webplayer" with podlove webplayer version 4 (standrad include)
- fixed opus file and other minimalistics
- create share rendering function in feed.php
- new endpoint for serach requests
- new endpoint to send sharing request
- fixed weblayer css and positions

RC 2.0

1.5 so far:
cleaned structure:
- lib, pwp, js, dict and classes moved to src-folder

changed way templating works:
- main template-folder contains default template (now quorx!)
- additional folders inside template-folder can contain additional files, overriding files inside default folder
- feed.cfg needs template: attribute
- feed.cfg can contain template-vars (go to @templatevars)
- template.cfg inside template-folder can contain default template-vars

1.3 update to podlove webplayer v2.0.13 and support images in mp4chaps

1.2 enabled rfc5005 inside standard feed (until now only by special urls...)
- fixed onother pile of smelly bugs. Take some care though...

1.1 another tons of bugfixes, templates per feed and podlove webplayer 2.0.11
- updated to fatfree 3.0.9, added adnthreads for episode comments

RC 1.0

tons of bugs fixed. podlove webplayer version 2.0.7 integrated

0.13 changed path to pages (now /show/page/@page from /page/@page) and pagers (now /show/pager/@pagenum from /show/page/@pagenum)
- added subdirs for /templates/pages: create dropdowns. only one level
- changed to podlove web player 2.0.5

0.12 fixed bug in markdown parsing (thanks bcosca)
- first (stable? haha) cloning support. call index.php from commandline

0.11 fixed pwp integration bug

0.10 changed to podlove webplayer v2.0.4

0.9 added bitlove support; reintroduced dropdown menus; articles now marked up by markdown

0.8 tons of fixes, typos and stuff. content-extensions.

0.7 autodetect mediafiles with mediabaseurl/path and formats:attribute

0.6 finished first draft of documentation. see doc/firtz.md

0.5 added subtemplate support for head/header/episode/footer
- fixed a bug, in case episodes didn't provide all configured media

0.4 added local auphonic support, yet work in progress
- singlefeed pages redirect to web page on / instead if listing one feed on frontpage

0.3 added first draft of extension class. atm only extends output templates (example: atom-extension. look into ext/atom/)
- changed website URL from /show/$feed to $feed/show to keep consistancy

0.2 major and minor fuckups fixed. thanks to all contributors!

0.1 let the world take a look


firtz-2.9/src/dict/de.php
<?php
return array(
    //search
    'dict_searchtext' => 'Suche...',

    //breadcrumb
    'dict_navi_allepisodes' => 'Alle Episoden',
    'dict_navi_episode' => 'Episode',
    'dict_navi_page' => 'Seite',

    //loop
    'dict_learnmore' => 'Mehr erfahren',
    'dict_readmore' => 'Reinhören',
    'dict_older' => 'Älter',
    'dict_newer' => 'Zurück',

    //contributors
    'dict_guest' => 'Gast',
    'dict_podcaster' => 'Podcaster',
    'dict_contris' => 'Mitwirkende',

    //episode meta
    'dict_publicBy' => 'Veröffentlicht von ',
    'dict_on' => 'am',
    'dict_epi_duration' => 'Diese Episode hat eine <b>Gesamtdauer</b> von:',
    'dict_epi_duration_end' => 'und steht Dir in den folgenden Formaten zum kostenlosen Download bereit:',

    //more
    'dict_keywords' => 'Stichwörter',
    'dict_subscribe' => 'Abonnieren',
    'dict_pubdate' => '{0,date}',
    'dict_duration' => 'Dauer',
    'dict_showcomments' => 'Kommentare',
    'dict_reply' => 'antworten',
    
    //license
    'dict_license' => 'Lizenz',
    'dict_lic_info' => 'Beachten Sie die Lizenzbestimmungen dieses Podcasts bevor Sie Inhalte weitergeben, einbinden oder gar verändern wollen. Die Einzelrechte an dieser Produktion unterliegen den jeweiligen Personen:',
    'dict_lic_end' => 'Produktion',
    'dict_epi_image' => 'Episoden Poster',
    'dict_epi_banner' => 'Episoden Banner',
    'dict_epi_intro' => 'Episoden Intro',
    'dict_epi_outro' => 'Episoden Outro',

    //sidebar
    'dict_sb_about_title' => 'Zum Podcast',
    'dict_sb_more' => 'Mehr Infos',
    'dict_sb_license' => 'Dieser Podcast unterliegt der ',
    
    'dict_sb_ultraschall_title' => 'Produziert mit Ultraschall',
    'dict_sb_ultraschall_p' => 'wird mit <a href="http://ultraschall.fm" title="Ultraschall Project">Ultraschall</a> erstellt. Dank des enorm vielseitigen und durchdachten PlugIns "Ultraschall" wird aus Reaper die Vorzeige-Software-Lösung für Podcastproduktionen, von der Aufnahme bis zum Mastering.',
    'dict_sb_podlove_title' => 'Podlove',
    'dict_sb_podlove_p' => '',

    'dict_sb_studiolink_title' => 'Studio&Link',
    'dict_sb_studiolink_p' => '',

);
?>


firtz-2.9/src/classes/episode.php
<?php

/**
 * Class for meta data from the episode
 *
 */
class episode
{
    public $item = array();
    public $main = "";
    public $markdown = "";

    /**
     * RC 3.0
     *
     * Episode constructor.
     *
     * @param $main
     * @param $ITEMFILE
     * @param $feedattrs
     * @param $slug
     * @param bool $auphonic
     * @param array $item
     */
    public function __construct($main, $ITEMFILE, $feedattrs, $slug, $auphonic = false, array $item = array())
    {
        if (!file_exists($ITEMFILE)) {
            $this->item = array();
            return;
        }
        $this->markdown = new Parsedown();
        $this->main = $main;
        $this->destroy = false;
        $reparse = false;

        if (is_array($item) && count($item) == 0) {
            // new item, set attributes
        } else {
            // reparsing an old one to overwrite data
            $reparse = true;
        }

        /*
         * Standard or Auphonic data (?)
         */
        if ($auphonic === false) {
            /* parse an .epi */
            $item = $this->parseConfig($main, $ITEMFILE, $feedattrs, $item);
        } else {
            /* parse the auphonic json file */
            $item = $this->parseAuphonic($main, $ITEMFILE, $feedattrs);
            if ($item === false) {
                $this->item = array();
                return;
            }
        }

        /*
         * just reparsing. skip the sanitation part
         */
        if ($reparse === true) {
            #$this->item = $item;
            #return;
        }

        /*
         * data sanitation
         */
        if ($item['date'] != "") {
            $pubDate = strtotime($item['date']);
            if ($pubDate === false) $pubDate = filectime($ITEMFILE);
        } else {
            $pubDate = filectime($ITEMFILE);
        }
        if ($pubDate > time()) {
            $this->item = array();
            $this->destroy = true;
            return;
        }

        /*
         * items meta data
         */
        $item['pubDate'] = date('D, d M Y H:i:s O', $pubDate);
        $item['pagelink'] = $main->get('BASEURL') . $feedattrs['slug'] . "/show/" . $slug;
        $item['slug'] = $slug;
        if ($item['guid'] == "") $item['guid'] = $feedattrs['slug'] . "-" . $item['slug'];
        $item['subtitle'] = substr(strip_tags($item['subtitle']), 0, 255); #new: subtitle / 3 RC

        // change: output on page in markdown and no strips
        // bug #1: find about gender* - https://github.com/Firtz-Designs/firtz3.0/issues/1
        $item['summary'] = substr($this->markdown->text($item['summary']),0,4000); #1 rc 3.0 - v2.9.1
        $item['summary_xml'] = substr(strip_tags($item['summary']), 0, 4000); #1 rc 3.0 - v2.9.1
        $item['shownotes_clean'] = $item['shownotes']; //editor clean
        $item['shownotes'] = $this->markdown->text($item['shownotes']);

        if ($item['image'] == "") $item['image'] = $feedattrs['image'];

        /*
         * who says, chapters are in order? sort them!
         */
        if ($item['chapters']) usort($item['chapters'], function ($a, $b) {
            return ($a['start'] > $b['start']);
        });

        if (strpos($item['duration'], '.') !== FALSE) $item['duration'] = substr($item['duration'], 0, strpos($item['duration'], '.'));

        $newkeywords = array();
        foreach (explode(",", $item['keywords']) as $key) $newkeywords[] = trim($key);
        $item['keywords'] = substr(implode(',', $newkeywords), 0, 255);
        $item['append'] = '';
        $item['prepend'] = '';

        // new: intro / 3 RC
        $itemIntro = explode(", ", $item['intro']);
        $item['intro_title'] = $itemIntro[0];
        $item['intro_owner'] = $itemIntro[1];
        $item['intro_ownerUrl'] = $itemIntro[2];
        $item['intro_lic'] = $itemIntro[3];
        $item['intro_licUrl'] = $itemIntro[4];

        $itemOutro = explode(", ", $item['outro']);
        $item['outro_title'] = $itemOutro[0];
        $item['outro_owner'] = $itemOutro[1];
        $item['outro_ownerUrl'] = $itemOutro[2];
        $item['outro_lic'] = $itemOutro[3];
        $item['outro_licUrl'] = $itemOutro[4];

        /*
         * get item
         */
        $this->item = $item;
    }

    /**
     * RC 3.0
     *
     * Parse standard epi file
     *
     * @param $main
     * @param $filename
     * @param $feedattrs
     * @param $item
     * @return array
     */
    public function parseConfig($main, $filename, $feedattrs, $item)
    {
        /*
         *  Parse an .epi file
         *
         * if item is given, it's a reparsing for overwriting
         * data from an auphonic-episode
         */
        $mime = $main->get('mimetypes');
        if (is_array($item) && count($item) == 0) {
            foreach ($main->get('itemattr') as $var) {
                $item = [$var];
            }
        }

        // vars
        $thisattr = array();
        $item['audiofiles'] = array();

        // file vars
        $slug = basename($filename, ".epi");
        $fh = fopen($filename, 'r');

        while (!feof($fh)) {
            $uline = fgets($fh);
            $line = trim($uline);

            /*
             * continue if comment or empty line. except for article attribute
             */
            if (($line == "" && $thisattr != "article") || substr($line, 0, 2) == "#:") continue;

            if ($line == "---end---") break;

            if (substr($line, -1) == ":" && (in_array(substr($line, 0, -1), $main->get('itemattr')) || in_array(substr($line, 0, -1), $feedattrs['audioformats']))) {

                // new attribute starts
                $thisattr = substr($line, 0, -1);

            } elseif ($thisattr == "chapters") {

                /*
                 * A chapter line
                 *
                 * no link or image atm will have to separate
                 * with or because of title attribute,
                 * which might contain white space
                 */
                preg_match('#((\d+:)?(\d\d?):(\d\d?)(?:\.(\d+))?) ([^<>\r\n]{3,}) ?(<([^<>\r\n]*)>\s*(<([^<>\r\n]*)>\s*)?)?\r?#', $line, $chapreg);
                $chap = array('start' => '', 'title' => '', 'image' => '', 'href' => '');
                if (isset($chapreg[1]) && isset($chapreg[6])) {
                    $chap['start'] = $chapreg[1];
                    $chap['title'] = trim($chapreg[6]);
                    if (isset($chapreg[8])) {
                        $chap['href'] = $chapreg[8];
                        if (isset($chapreg[10])) {
                            $chap['image'] = $chapreg[10];
                        }
                    }
                    $item['chapters'][] = (array)$chap;
                }

            } elseif (in_array($thisattr, $feedattrs['audioformats'])) {

                /*
                 * Configured audio formats
                 *
                 * only audioformats allowed,
                 * that are configured in feed.cfg
                 */
                $audio = explode(" ", $line);
                if (!array_key_exists($thisattr, $mime)) {
                    // mimetype not found in presets
                    if (count($audio) == 3) {
                        // maybe it's in the .epi?
                        $mimetype = $audio[2];
                    } else {
                        // fallback
                        $mimetype = "audio/mpeg";
                    }
                } else {
                    // everything went better than expected
                    $mimetype = $mime[$thisattr];
                }

                if (count($audio) > 1) {
                    $item[$thisattr] = array('link' => $audio[0], 'length' => $audio[1], 'type' => $mimetype);
                } else {
                    $item[$thisattr] = array('link' => $audio[0], 'length' => 0, 'type' => $mimetype);
                }

                $item['audiofiles'][$thisattr] = $item[$thisattr];

            } else {
                // this is an attribute which may have linebreaks. append line to current attribute
                if ($thisattr != "" && $thisattr != "article") $item[$thisattr] .= ($item[$thisattr] != "") ? "\n" . $line : $line;
                if ($thisattr == "article") $item[$thisattr] .= ($item[$thisattr] != "") ? "\n" . $uline : $uline;
            }

        }

        if ($feedattrs['mediabaseurl'] != "") {
            foreach ($feedattrs['audioformats'] as $format) {
                if ($feedattrs['mediabasepath'] != "") {
                    $localfile = $feedattrs['mediabasepath'] . "/" . $slug . "." . $format;
                    if (!file_exists($localfile)) continue;
                } else {
                    $localfile = "";
                }

                if (array_key_exists($format, $mime)) {
                    $mimetype = $mime[$format];
                } else {
                    $mimetype = "audio/mpeg";
                }

                if (isset($item['audiofiles'][$format])) {
                    $audiofilename = basename($item['audiofiles'][$format]['link'], "." . $format);
                } else {
                    $audiofilename = $slug;
                }
                $item[$format] = array('link' => $feedattrs['mediabaseurl'] . $audiofilename . "." . $format, 'length' => $localfile != "" ? filesize($localfile) : 0, 'type' => $mimetype);
                $item['audiofiles'][$format] = $item[$format];
            }
        }

        fclose($fh);
        return $item;
    }

    /**
     * RC. 2.1 (please test)
     *
     * Parse Auphonic json
     *
     * @param $main
     * @param $filename
     * @param $feedattrs
     * @return bool
     */
    public function parseAuphonic($main, $filename, $feedattrs)
    {
        /*
         * parse a json production description file
         */
        $mime = $main->get('mimetypes');
        foreach ($main->get('itemattr') as $var) $item[$var] = "";

        $thisattr = "";

        $prod = json_decode(str_replace("\\r", "\\n", file_get_contents($filename)));

        if ($prod === false) return false;

        $item['title'] = $prod->metadata->title;
        $item['description'] = strip_tags($prod->metadata->subtitle);
        $item['article'] = $prod->metadata->summary;

        if (isset($prod->metadata->location)) {
            if (isset($prod->metadata->location->latitude) && isset($prod->metadata->location->longitude)) {
                $item['location']['latitude'] = $prod->metadata->location->latitude;
                $item['location']['longitude'] = $prod->metadata->location->longitude;
            }
        }

        $item['duration'] = $prod->length_timestring;
        $item['date'] = date('r', strtotime($prod->change_time));

        // images in chapters are located on auphonics server.
        foreach ($prod->chapters as $chapter) {
            $chap = array('start' => $chapter->start, 'title' => $chapter->title, 'image' => '', 'href' => '');
            if (isset($chapter->url)) $chap['href'] = $chapter->url;
            // if (isset($chapter->image)) $chap['image'] = $chapter->image;
            $item['chapters'][] = $chap;
        }
        if (isset($prod->multi_input_files) && $item['chapters'] != "") {
            foreach ($prod->multi_input_files as $mif) {
                if ($mif->type == 'intro') {
                    foreach ($item['chapters'] as $key => $chap) $item['chapters'][$key]['start'] = strftime("%H:%M:%S", strtotime($chap['start']) + $mif->input_length);
                }
            }
        }

        $services = array();
        foreach ($prod->outgoing_services as $service) {
            // only services with a base_url work...
            if (isset($service->base_url) && $service->base_url != "") {
                if (substr($service->base_url, -1) == '/')
                    $services[$service->uuid] = $service->base_url;
                else
                    $services[$service->uuid] = $service->base_url . '/';
            }
            if ($service->type == "amazons3") $services[$service->uuid] = 'http://' . $service->bucket . '.s3.amazonaws.com/';
        }
        $item['audiofiles'] = array();

        foreach ($prod->output_files as $output) {
            if (count($output->outgoing_services) == 0) continue;

            $service = "";

            // Check if this services exists
            foreach ($output->outgoing_services as $oservice) {
                if (array_key_exists($oservice, $services)) {
                    $service = $oservice;
                    break;
                }
            }

            if ($service == "") continue;
            if ($output->format == "image") $item['image'] = $services[$service] . $output->filename;
            if (in_array($output->ending, $feedattrs['audioformats'])) {
                $mimetype = (array_key_exists($output->ending, $mime) ? $mime[$output->ending] : "application/octet");
                $item[$output->ending] = array('link' => $services[$service] . $output->filename, 'length' => $output->size, 'type' => $mimetype);
                $item['audiofiles'][$output->ending] = $item[$output->ending];
            }
        }

        /*
         * Replace attributes by tags,
         * starting with eg:
         *
         *   date: 2013-12-17 10:00:00
         */
        foreach ($prod->metadata->tags as $key => $tag) {
            $tag = trim($tag);
            if (substr($tag, 0, 1) == "_" && strpos($tag, ':') !== false) {
                $tagname = substr($tag, 1, strpos($tag, ':') - 1);
                $tagval = trim(substr($tag, strpos($tag, ':') + 1));

                if (in_array($tagname, $main->get('itemattr'))) {
                    $item[$tagname] = $tagval;
                    unset($prod->metadata->tags[$key]);
                }
            }
        }

        $item['keywords'] = implode(",", $prod->metadata->tags);
        return $item;
    }

    /**
     * RC 0.1
     *
     * This function is obsolete and
     * kept for nostalgic reasons
     *
     * @param $audioformat
     * @return string
     */
    public function renderRSS2($audioformat)
    {
        $main = $this->main;
        $this->item['enclosure'] = $this->item[$audioformat];
        $main->set('item', $this->item);
        return Template::instance()->render('rss2_item.xml', 'application/xml');
    }

    /**
     * RC 0.1
     *
     * This function is obsolete
     * and kept for nostalgic reasons
     */
    public function renderHTML()
    {
        $main = $this->main;
        $main->set('item', $this->item);
        echo Template::instance()->render('html_item.html');
    }

}
?>